convert_zdf.py

```diff
@@ -0,0 +1,168 @@
+"""Convert ZDF point cloud to the available formats.
+
+Available formats:
+    .ply - Polygon File Format
+    .csv,.txt - [X, Y, Z, r, g, b, Contrast]
+    .png,.jpg,.bmp,.tiff - 2D RGB image
+"""
+
+import argparse
+import os
+import numpy as np
+import cv2
+import zivid
+
+
+def _options():
+    parser = argparse.ArgumentParser(
+        description="Convert from ZDF to your preferred format",
+        epilog="Example: $ python -u convert_zdf.py --csv --png Zivid3D.zdf Zivid3D_1.zdf",
+    )
+
+    parser.add_argument(
+        "files", nargs="*", help="List of files to convert", default="Zivid3D.zdf"
+    )
+    parser.add_argument("--ply", action="store_true", help="Convert to PLY")
+    parser.add_argument("--csv", action="store_true", help="Convert to CSV")
+    parser.add_argument("--txt", action="store_true", help="Convert to text")
+    parser.add_argument(
+        "--png",
+        action="store_true",
+        help="Convert to Portable Network Graphics (3D->2D)",
+    )
+    parser.add_argument("--jpg", action="store_true", help="Convert to JPEG (3D->2D)")
+    parser.add_argument(
+        "--bmp", action="store_true", help="Convert to Windows bitmap (3D->2D)"
+    )
+    parser.add_argument("--tiff", action="store_true", help="Convert to TIFF (3D->2D)")
+
+    return parser.parse_args()
+
+
+def _flatten_point_cloud(point_cloud):
+    """Convert from point cloud to flattened point cloud (with numpy).
+
+    Args:
+        point_cloud: a numpy array
+
+    Returns:
+        2D numpy array, with 7 columns and npixels rows
+
+    """
+    # Convert to numpy 3D array
+    point_cloud = np.dstack(
+        [
+            point_cloud["x"],
+            point_cloud["y"],
+            point_cloud["z"],
+            point_cloud["r"],
+            point_cloud["g"],
+            point_cloud["b"],
+            point_cloud["contrast"],
+        ]
+    )
+    # Flattening the point cloud
+    flattened_point_cloud = point_cloud.reshape(-1, 7)
+    # Just the points without color and contrast
+    # point_cloud = np.dstack([point_cloud['x'],point_cloud['y'],point_cloud['z']])
+    # flattened_point_cloud = point_cloud.reshape(-1,3)
+
+    # Removing nans
+    return flattened_point_cloud[~np.isnan(flattened_point_cloud[:, 0]), :]
+
+
+def _convert_2_ply(frame, file_name):
+    """Convert from frame to ply.
+
+    Args:
+        frame: A frame captured by a Zivid camera.
+        file_name: File name without extension
+
+    """
+    print(f"Saving the frame to {file_name}.ply")
+    frame.save(file_name + ".ply")
+
+
+def _convert_2_csv(point_cloud, file_name):
+    """Convert from point cloud to csv.
+
+    Args:
+        point_cloud: a numpy array
+        file_name: File name without extension
+
+    """
+    print(f"Saving the frame to {file_name}.csv")
+    np.savetxt(
+        file_name + ".csv", _flatten_point_cloud(point_cloud), delimiter=",", fmt="%.3f"
+    )
+
+
+def _convert_2_txt(point_cloud, file_name):
```

Comment:
All of these functions look very much the same, could this have been shortened?

Reply:
Yes, I thought about refactoring as soon as I added more than one.. let me have a look. I still want the log to reflect which conversion is ongoing